src/Common/src/Internal/Text/Utf8String.cs

```diff
@@ -35,5 +38,20 @@ public override string ToString()
         {
             return Encoding.UTF8.GetString(_value);
         }
+
+        public override bool Equals(object obj)
+        {
+            return (obj is Utf8String) && Equals((Utf8String)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return ((IStructuralEquatable)_value).GetHashCode(EqualityComparer<byte>.Default);
```

Comment:
I kind of knew you wouldn't let me be lazy about this.

I committed the update as you, since you pretty much wrote it all anyway (minus a few bug fixes).